file parsing solution

kbrose opened the pull request:
### Overview

**This PR is a work in progress.**

This version of the code does a rudimentary parsing of the python source code, keeping track of brackets and doing **_very**_ rudimentary bracket matching (if the python source code is not well-formed then bad things could happen).
### Implementation notes

(A note on terminology, I use "bracket" to refer to any of the `[](){}` characters.)

Essentially, simple parsing is performed to keep a stack of the column-location of open brackets (one of `[({`). When an opening bracket is read, it adds to the stack indicating the column where the bracket is located, when a closing bracket (one of `})]`) is read, it pops the latest element (it does not do any checking to make sure the closing bracket matches the opening bracket - this is why it assumes the python source code is well-formed). If there is anything in the stack after parsing the python file up to the cursor location, then that means there is an open bracket. We can see what column the most recent addition to the stack was on, and then set the indent from that.
### TODO

Ordered in what I perceive to be most important (top) to least important (bottom):
- Should intelligently un-indent back to the correct level after all brackets have been closed.
- All of the settings and things are still in the file. They will need to be updated (to only include the amount of the hanging indent?)
- Only supports soft-tabs. It seems to me that the pep8 convention does not even make sense with hard-tabs, since it explicitly requires being able to have preceding whitespace of any amount, not necessarily divisible by the length of the tab. Still, though, it would be nice to support this. Perhaps until we can do it, we should return without any processing if hard-tabs are being used. Looking at https://github.com/atom/atom/blob/master/src/text-editor.coffee#L2702 suggests that you can build an indent string of arbitrary length even when hard-tabs are being used (it just pads with spaces after using as many tabs as possible).
- I added a python file for testing (`test/test_file.py`) that has all of the more pathological cases I could think of. Is it possible to automate a test based on this somehow? Don't really know much about testing or testing frameworks.
- Style-guide and documentation.
### Finished
- This implementation removed the hanging indent feature. Could probably detect the hanging indent the same way as before with the regex. Just need to add it and try it.

kbrose committed:
Have parseLine and parseLines functions

Have not tested at all, no idea if it will even "compile".

kbrose committed:
Fix iterator stop condition and recursion error

kbrose committed:
parsing working for square brackets and parens

kbrose committed:
simplified stack, handles curly braces

kbrose committed:
adding TODOs

kbrose committed:
cleaning up code, adding some comments

kbrose committed:
adding a python file to test on

kbrose committed:
realized tracking the row was not needed

kbrose committed:
adding hanging indent

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -42,31 +47,29 @@ module.exports = PythonIndent =
 
     # If there's a remainder, `editor.buildIndentString` requires the tab to
     # be set past the desired indentation level, thus the ceiling.
+    # TODO: is the conditional necessary? Maybe to avoid floating point errors?
```
Yes, this is because of the possibility that `rem` could be 0, which is totally acceptable. I remember this when testing through everything, though I can't remember the exact scenario.

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -42,31 +47,29 @@ module.exports = PythonIndent =
+    col = editor.getCursorBufferPosition().column
+
+    # Parse the entire file up to the current point, keeping track of brackets
+    lines = editor.getTextInBufferRange([[0,0], [row, col]]).split('\n')
```
I do think of potential performance issues on very large (> 100 lines) files. I think we'll want to put some load with lots of brackets to test potentially parsing the whole file.

kbrose replied:
Good idea. I figure that compared to the parsing the`language-python` package is doing in order to get syntax highlighting correct this would be minimal. Should still check that it has good performance though.

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -75,32 +78,58 @@ module.exports = PythonIndent =
+    # The range of text to replace with our indent
+    # will need to change this for hard tabs, especially tricky for when
+    # hard tabs have mixture of tabs + spaces, which they can judging from
+    # the editor.buildIndentString function
```
Yes, that function allows for a mixture of hard tabs and spaces, but PEP8 does not, if I remember correctly. The interpreter might even fail when tabs and spaces are mixed.

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -75,32 +78,58 @@ module.exports = PythonIndent =
+                    bracketStack.push([row, col])
+                if c in '})]'
+                    bracketStack.pop()
+                if c in '\'"'
```
Nice...check if in a quoted string.

DSpeckhals replied:
Also, perhaps it would be better to put this in an `if...else if` control structure. That way this can be short-circuited once a match is found.

I actually saw similar logic to this in a Sublime Text plugin once.

kbrose replied:
Oh, yeah good catch. I originally had it as a switch but then changed it out for `if/then`, but you're totally right to `else`s in there.

DSpeckhals commented on test/test_file.py:
```diff
@@ -0,0 +1,45 @@
+# This file represents all known cases of special-case indentations
```
Good idea. I like the idea of having a "crazy scenario" file like this. I actually have one, but it's ignored in the repo :smile:

kbrose replied:
Yep, found it invaluable for testing. But now that I've had the CI tests fail I see that there's sort of one in the `specs` folder. I'll have to look into that more (also to see why the tests are failing...)

DSpeckhals replied:
I think it was mainly hanging indent failures. They're Jasmine tests, which
are rather simple and easy to get started with. I only wrote my first a few
months ago.

Typically, you don't want too many outside file dependencies for a unit
test, but for this case, I wouldn't see a problem with loading it into an
array split on lines, then test using that.

DSpeckhals commented:
I made some basic comments on what you have here. Excellent work so far. These kinds of improvements would make the package much better. Thanks!

Edit: I was making some changes/major refactors in the past hour or so. I might merge them this week. They don't really change logic...just module structure/organization. Just keep on working with what you've got, and I can always fix any conflicts that come from your branch later.

kbrose commented:
Sounds good.

kbrose committed:
Adding else clauses to short-circuit parsing

Per the suggestion by DSpeckhals.

kbrose committed:
Correctly reset indent level after closing bracket

Sort of works with function definitions as well. More to come on that
later.

kbrose committed:
Hanging indent fix broke nested list continuations

kbrose committed:
correctly indent hanging indents and functions

kbrose committed:
cleaning code

kbrose commented:
Ok, as of db36a0b this is working on all of the test cases (which are all in `test.py`) I could think of (let me know if I missed any. This was on Windows 10. Whenever you've got the time, could you look it over and make comments on what you want changed? This is my first time writing any JS/coffeescript, so style comments are definitely welcome in addition to practical things.

BTW, I also had a ~600 line python file that I played around with and noticed no performance issues.

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -42,28 +30,62 @@ module.exports = PythonIndent =
+    lastFunctionRow = parseOutput.lastFunctionRow
+
+    if shouldHang
+        console.log('here')
```
As you can probably guess, you'll want to take out the log statements.

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -42,28 +30,62 @@ module.exports = PythonIndent =
+    lines = editor.getTextInBufferRange([[0,0], [row, col]]).split('\n')
+    # at this point, the newline character has just been added,
+    # so remove the last element of lines, which will be the empty line
+    lines = lines.splice(0,lines.length-1)
```
This should be spaced like `lines = lines.splice(0, lines.length - 1)`.

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -42,28 +30,62 @@ module.exports = PythonIndent =
+    if not openBracketStack.length
+        # can assume closeBracketStack is not empty
+        lastClosedBracketLocations = closeBracketStack.pop()
+        if lastClosedBracketLocations[1] == row-1
```
In Coffeescript, any `==` should be replaced by either `is` or `===`. Both of these check for equality in type and value.

kbrose replied:
Trying out some code on [http://coffeescript.org/](http://coffeescript.org/) seems to suggest that `==` and `is` in coffeescript both get mapped directly to `===` in JavaScript, and `===` is an error.

DSpeckhals replied:
Ahhh...well I learned something new. I typically use `is`, so I guess
that's why I haven't seen errors :smile:

Dustin Speckhals
On Feb 25, 2016 8:43 PM, "Kevin Rose" notifications@github.com wrote:

> In lib/python-indent.coffee
> https://github.com/DSpeckhals/python-indent/pull/9#discussion_r54195112:
> 
> > -    openBracketStack = parseOutput.openBracketStack
> > -    closeBracketStack = parseOutput.closeBracketStack
> > -    shouldHang = parseOutput.shouldHang
> > -    lastFunctionRow = parseOutput.lastFunctionRow
> >   +
> > -    if shouldHang
> > -        console.log('here')
> > -        PythonIndent.indentHanging(editor, row, editor.buffer.lineForRow(row - 1))
> > -        return
> >   +
> > -    return unless openBracketStack.length or closeBracketStack.length
> >   +
> > -    if not openBracketStack.length
> > -        # can assume closeBracketStack is not empty
> > -        lastClosedBracketLocations = closeBracketStack.pop()
> > -        if lastClosedBracketLocations[1] == row-1
> 
> Trying out some code on http://coffeescript.org/ seems to suggest that ==
> and is in coffeescript both get mapped directly to === in JavaScript, and
> === is an error.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/DSpeckhals/python-indent/pull/9/files#r54195112.

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -42,28 +30,62 @@ module.exports = PythonIndent =
+                # we just finished defining a function, need to increase indentation
+                indentLevel += 1
+
+            editor.setIndentationForBufferRow(row, indentLevel)
```
I'm not that picky about it, but in Coffeescript, you can say this as `editor.setIndentationForBufferRow row, indentLevel`

DSpeckhals commented on lib/python-indent.coffee:
```diff
@@ -75,28 +97,87 @@ module.exports = PythonIndent =
+    output.closeBracketStack = closeBracketStack
+    output.shouldHang = shouldHang
+    output.lastFunctionRow = lastFunctionRow
+    return output
```
That would work, but I might say this instead:

``` coffeescript
return {} =
    openBracketStack: openBracketStack
    closeBracketStack: closeBracketStack
    shouldHang: shouldHang
    lastFunctionRow: lastFunctionRow
```

DSpeckhals commented:
I left some more comments. This is getting pretty close! When you're happy with me testing this for a day or 2 in real life, let me know; then I'll pull your branch, resolve conflicts, and try it out.

kbrose committed:
Documentation, style, and cleaning

Added documentation, fixed some style concerns. Only keep track of the
last place a bracket was closed instead of a whole stack since that's all
that's needed for the parsing. Rearranged logic for when hanging indent
should be used to accomodate comments on the line before a hanging indent
is used.

kbrose committed:
adding comment before hanging indent test to file

kbrose commented:
Added more documentation to the parsing section, fixed the style suggestions, and made a couple other minor changes (one performance based, the other should allow hanging indents to be correctly triggered when a comment was entered before pressing enter for the newline that should have the hanging indent). I believe this version (cd8814a) should be ready for you to test.

DSpeckhals commented:
Super! I'll hammer away at it tomorrow.

DSpeckhals commented:
I resolved the conflicts without much trouble, and will test over the next 24 hours or so. I should be able to release tomorrow.

The reason I originally put this package together was for purely selfish reasons. I was surprised by the response and help from everyone. Your contributions have made it magnitudes better. I look forward to using it "out in the wild." Thanks so much, again! :clap:
